File: Assets/Bundles/Server/Resources/TS/Services/Global/Chat/Commands/Generator/CreateGeneratorCommand.lua
-- Compiled with unity-ts v2.1.0-75
local Flamework = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Flamework
local ItemType = require("Shared/TS/Item/ItemType").ItemType
local ChatCommand = require("Server/TS/Services/Global/Chat/Commands/ChatCommand").ChatCommand
-- * Default generator item type.
local DEFAULT_ITEM_TYPE = ItemType.COBBLESTONE
-- * Default generator spawn rate.
local DEFAULT_SPAWN_RATE = 2
-- * Default generator stack limit.
local DEFAULT_GENERATOR_STACK_LIMIT = 50
local CreateGeneratorCommand
do
	local super = ChatCommand
	CreateGeneratorCommand = setmetatable({}, {
		__tostring = function()
			return "CreateGeneratorCommand"
		end,
		__index = super,
	})
	CreateGeneratorCommand.__index = CreateGeneratorCommand
	function CreateGeneratorCommand.new(...)
		local self = setmetatable({}, CreateGeneratorCommand)
		return self:constructor(...) or self
	end
	function CreateGeneratorCommand:constructor()
		super.constructor(self, "createGenerator")
	end
	function CreateGeneratorCommand:Execute(player, args)
		local itemType
		local spawnRate
		-- If no arguments are provided fallback to defaults.
		if #args == 0 then
			itemType = DEFAULT_ITEM_TYPE
			spawnRate = DEFAULT_SPAWN_RATE
		end
		-- ItemType argument provided, fallback to default spawn rate.
		if #args == 1 then
			itemType = args[1]
			spawnRate = DEFAULT_SPAWN_RATE
		end
		-- ItemType and spawn rate arguments provided.
		if #args == 2 then
			itemType = args[1]
			spawnRate = tonumber(args[2])
		end
		-- If no ItemType or spawn rate, return.
		if itemType == nil or spawnRate == nil then
			player:SendMessage("Invalid arguments")
			return nil
		end
		-- Spawn generator underneath command executor.
		local executorEntity = (Flamework.resolveDependency("Bundles/Server/Services/Global/Entity/EntityService@EntityService")):GetEntityByClientId(player.clientId)
		if not executorEntity then
			return nil
		end
		local generatorPosition = executorEntity.gameObject.transform.position;
		(Flamework.resolveDependency("Bundles/Server/Services/Global/Generator/GeneratorService@GeneratorService")):CreateGenerator(generatorPosition, {
			item = itemType,
			spawnRate = spawnRate,
			stackLimit = DEFAULT_GENERATOR_STACK_LIMIT,
			label = true,
		})
	end
end
return {
	CreateGeneratorCommand = CreateGeneratorCommand,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Server/Services/Global/Chat/Commands/Generator/CreateGeneratorCommand.ts
import { Dependency } from "@easy-games/flamework-core";
import { ChatCommand } from "Imports/Core/Shared/Commands/ChatCommand";
import { EntityService } from "Server/Services/Global/Entity/EntityService";
import { GeneratorService } from "Server/Services/Global/Generator/GeneratorService";
import { ItemType } from "Shared/Item/ItemType";
import { Player } from "Shared/Player/Player";

/** Default generator item type. */
const DEFAULT_ITEM_TYPE = ItemType.COBBLESTONE;
/** Default generator spawn rate. */
const DEFAULT_SPAWN_RATE = 2;
/** Default generator stack limit. */
const DEFAULT_GENERATOR_STACK_LIMIT = 50;

export class CreateGeneratorCommand extends ChatCommand {
	constructor() {
		super("createGenerator");
	}

	public Execute(player: Player, args: string[]): void {
		let itemType: ItemType | undefined;
		let spawnRate: number | undefined;

		/* If no arguments are provided fallback to defaults. */
		if (args.size() === 0) {
			itemType = DEFAULT_ITEM_TYPE;
			spawnRate = DEFAULT_SPAWN_RATE;
		}

		/* ItemType argument provided, fallback to default spawn rate. */
		if (args.size() === 1) {
			itemType = args[0] as ItemType;
			spawnRate = DEFAULT_SPAWN_RATE;
		}

		/* ItemType and spawn rate arguments provided. */
		if (args.size() === 2) {
			itemType = args[0] as ItemType;
			spawnRate = tonumber(args[1]);
		}

		/* If no ItemType or spawn rate, return. */
		if (itemType === undefined || spawnRate === undefined) {
			player.SendMessage("Invalid arguments");
			return;
		}

		/* Spawn generator underneath command executor. */
		const executorEntity = Dependency<EntityService>().GetEntityByClientId(player.clientId);
		if (!executorEntity) return;
		const generatorPosition = executorEntity.gameObject.transform.position;
		Dependency<GeneratorService>().CreateGenerator(generatorPosition, {
			item: itemType,
			spawnRate: spawnRate,
			stackLimit: DEFAULT_GENERATOR_STACK_LIMIT,
			label: true,
			/* TODO: Split data? Maybe? */
		});
	}
}
